os/pm/pm_timer/pm_timer_activity.c

```diff
@@ -0,0 +1,244 @@
+		}
+
+		if (PM_ISSTATIC(head)) {
+			/* Put the pm timer back on the free list */
+			sq_addlast((FAR sq_entry_t *)head, &g_pmTimer_freeList);
+			g_pmTimer_nfree++;
+			DEBUGASSERT(g_pmTimer_nfree <= CONFIG_PM_MAX_WAKEUP_TIMER);
+			leave_critical_section(state);
+
+		} else if (PM_ISALLOCED(head)) {
+			/* Free the dynamically allocated timer 
+			 * If the timer was released from an interrupt handler, 
+			 * sched_kfree() will defer the actual deallocation of the
+			 * memory until a more appropriate time. 
+			 * 
+			 * We don't need interrupts disabled to do this */
+			leave_critical_section(state);
+			sched_kfree(head);	
+
+		} else {
+			/* This should not be the case because a pm timer 
+			 * is either statically allocated or dynamically */
+			pmdbg("This should not be any case\n");
+			leave_critical_section(state);
+		}
+	}
+}
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/****************************************************************************
+ * Name: pm_set_wakeup_timer
+ *
+ * Description:
+ *   This function is called just before sleep to start the required PM wake up
+ *   timer. It will start the first timer from the g_pmTimer_activeList with the
+ *   required delay.(delay should be positive)
+ * 
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   0 - system can go to sleep
+ *   -1 - system should not go to sleep
+ *
+ ****************************************************************************/
+
+int pm_set_wakeup_timer(void)
+{
+	/* if no timer present , do we need to wake up board ??? */
+	if (g_pmTimer_activeList.head == NULL) {
+		pmdbg("the g_pmTimer_activeList list is empty!!!\n");
+		return PM_TIMER_SUCCESS;
+	}
+
+	pm_wakeup_timer_t *curr;
+    	pm_wakeup_timer_t *prev;
+    	prev = curr = (pm_wakeup_timer_t *)g_pmTimer_activeList.head;
+
+	/* Remove the wakeup timers that are already expired. 
+	 * Although this must have already handled by pm_timer_update() */
+	while (curr->delay <= 0 && curr->next) {
+		prev = curr;
+		curr = curr->next;
+	}
+
+	if (curr->delay > 0) {
+		pmdbg("Setting timer and Board will wake up after %d millisecond\n", curr->delay);
+		pm_set_timer(PM_WAKEUP_TIMER, curr->delay * 1000);
+	} 
+	
+	return PM_TIMER_SUCCESS;
+}
+
+/****************************************************************************
+ * Name: pm_timer_update
+ *
+ * Description:
+ *   This function decreases the delay of head pm timer in the 
+ *   g_pmTimer_activeList by 1 after every sys tick. If the delay becomes 0,
+ *   It expires the pm timer.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None.
+ * 
+ * Assumption: This should be also implemented for CONFIG_SCHED_TICKLESS and
+ * CONFIG_SCHED_TICKSUPRESS.
+ *
+ ****************************************************************************/
+
+void pm_timer_update(void)
```

Comment:
we need this function for only -- tick?

How about use below fucntion pm_wakeup_handler use only with argument tick=1?